### DISCUSSION TITLE
More iterator methods (any, all, windows, chunks, take, etc.)

### DISCUSSION THREAD

**Discussion by mcastorina** [vlang/v#11814] [Ideas]
`filter` and `map` are great but I'd like more iterator methods like the ones mentioned in the title. Rust offers a huge number of functions for anything that is iterable: https://doc.rust-lang.org/std/iter/trait.Iterator.html#provided-methods

The biggest benefit I see is it allows for a more functional style of programming.

Some example usage:
```v
x := [1, 2, 3]
x.any(it == 2) // true
x.all(it < 2)  // false
x.sum()        // 6
x.chunks(2)    // [[1, 2], [3]]
x.windows(2)   // [[1, 2], [2, 3]]
x.take(2)      // [1, 2]
x.skip(1)      // [2, 3]
```

**Comment by mcastorina** [comment]
I'm still new to V but I would be happy to work on these features if approved by the community / maintainers.

**Comment by SleepyRoy** [comment]
Hi @mcastorina yep V embraces FP style and want to hear more about it! I suggest you join V's discord channel (find it on main page). There you can share your idea and have a direct and interactive discussion with community and many dev members. IMO it's an effective way to get approval of a proposal : )

**Comment by Delta456** [comment]
`.any` and `.all` have been implemented in https://github.com/vlang/v/pull/9347

**Comment by nordlow** [comment]
Please take into account D's corresponding concept *ranges*.

https://tour.dlang.org/tour/en/basics/ranges
https://dlang.org/phobos/std_range.html
https://wiki.dlang.org/Component_programming_with_ranges

A D range is closely related to a Rust iterator but not to a C++ iterator. In D a range is a combination of a C++ begin and end iterator.

Like in C++ and Rust ranges are not limited to array operations. They instead return lazy and chainable calculations. D ranges have a couple of powerful concepts not present in most other languages. For instance the return type of `std.algorithm.iteration.map` is a `RandomAccessRange` iff its input range is a `RandomAccessRange.` Meaning you can sort elements from the map call in-place. The same goes for `std.range.chain`. If all it's input ranges are `RandomAccessRange` so is the returned chained range. That means that you can sort a set of disjunct memory region in-place if you feel like.

Moreover ranges can be queried at compile-time for unboundedness (empty property is always is false). In D such ranges as called infinite ranges or generators.
